File: 2021/dec05/joakim_loefgren/day5.jl
function parse_input(input_file="./input_day5.txt")
    lines = open(input_file, "r") do f
        readlines(f)
    end
    re = r"(\d+)"
    vents = zeros(Int64, 4, length(lines))
    for (i, line) in enumerate(lines)
        nums = parse.(Int, SubString.(line, findall(re, line))) .+ 1
        if nums[1] > nums[3]
            nums[1], nums[3] = nums[3], nums[1]
            nums[2], nums[4] = nums[4], nums[2]
        end
        vents[:, i] = nums
    end
    return vents
end


function build_seafloor(vents::Array{Int64, 2}, diagonal::Bool=false)
    nx = max(maximum(@view vents[1, :]), maximum(@view vents[3, :]))
    ny = max(maximum(@view vents[2, :]), maximum(@view vents[4, :]))
    seafloor = zeros(Int, ny, nx)
    nvents = size(vents)[2]
    for n in 1:nvents
        if vents[1, n] == vents[3, n]  # vertical
            jmin = min(vents[2, n], vents[4, n]) 
            jmax = max(vents[2, n], vents[4, n]) 
            seafloor[jmin : jmax, vents[1, n]] .+= 1
        elseif vents[2, n] == vents[4, n]  # horizontal
            seafloor[vents[2, n], vents[1, n] : vents[3, n]] .+= 1
        else
            if diagonal == true  # diagonal
                if vents[4, n] > vents[2, n]
                    for k in 0:(vents[3, n] - vents[1, n])
                        seafloor[vents[2, n] + k, vents[1, n] + k] += 1
                    end
                else
                    for k in 0:(vents[3, n] - vents[1, n])
                        seafloor[vents[2, n] - k, vents[1, n] + k] += 1
                    end
                end
            end
        end
    end
    return seafloor
end

function main() 
    # Part I
    vents = parse_input()
    seafloor = build_seafloor(vents)
    println(sum(seafloor .>= 2))

    # Part II
    seafloor = build_seafloor(vents, true)
    println(sum(seafloor .>= 2))
end

main()

File: 2021/dec05/carl_toft/day5.py
import numpy as np
from utils import read_lines

def parseInput(filename):
    """Parse input and return it as a 2D array."""
    input = read_lines(filename)
    lines = np.zeros((len(input), 4), dtype=int)
    for idx, line in enumerate(input):
        start, end = line.split('->')
        start = start.strip()
        end = end.strip()
        start = tuple((int(x) for x in start.split(',')))
        end = tuple((int(x) for x in end.split(',')))
        lines[idx, 0] = start[0]
        lines[idx, 1] = start[1]
        lines[idx, 2] = end[0]
        lines[idx, 3] = end[1]
    return lines

lines = parseInput("/home/carl/Code/AdventOfCode/Day5/input.txt")

# Build a map that fits all lines
size = np.max(lines)+1
map = np.zeros((size, size))

# Add all lines to the map
for k in range(lines.shape[0]):
    # Only look at horizontal or vertical lines
    if lines[k,0] != lines[k,2] and lines[k,1] != lines[k,3]:
        # The line is diagonal. Parametrize it by the x-values
        if lines[k,2] > lines[k,0]:
            x_start = lines[k,0]
            x_stop = lines[k,2]+1
            y_start = lines[k,1]
            y_stop = lines[k,3]+1
        else:
            x_start = lines[k,2]
            x_stop = lines[k,0]+1
            y_start = lines[k,3]
            y_stop = lines[k,1]+1
        y_step = np.sign(y_stop - y_start)
        for kk in range(x_stop-x_start):
            map[y_start+y_step*kk, x_start + kk] += 1

    if lines[k,0] == lines[k,2]:
        # The line is vertical
        map[lines[k,1]:lines[k,3]+1, lines[k,0]] += 1
        map[lines[k,3]:lines[k,1]+1, lines[k, 0]] += 1
    if lines[k,1] == lines[k,3]:
        # The line is horizontal
        map[lines[k,1], lines[k,0]:lines[k,2]+1] += 1
        map[lines[k,1], lines[k,2]:lines[k,0] + 1] += 1

print(map)
print('Part 2:', np.sum(map>=2))
